docs/standard/native-interop/best-practices.md

```diff
@@ -251,6 +251,58 @@ finally
 }
 ```
 
+## Windows and non-Windows data type considerations
+
+There are types in the C language that have latitude in how they are defined. When writing cross-platform interop, cases can arise where platforms differ and can cause issues if not considered.
+
+### C `long`
+
+The `long` in C is defined to have ["at least 32"](https://en.cppreference.com/w/c/language/arithmetic_types) bits. This means there is a minimum required but platforms are not limited to that. The table below illustrates the differences in provided bits for the C `long` data type between platforms.
+
+|             | 32-bit | 64-bit |
+|:------------|:-------|:-------|
+| Windows     | 32     | 32     |
+| macOS/\*nix | 32     | 64     |
+
+These differences make authoring cross-platform P/Invokes difficult when the native function is defined to use `long` on all platforms. When targeting .NET 5 and prior .NET versions, you should declare separate Windows and non-Windows signatures to handle the problem.
+
+```csharp
+// Cross platform C function
+// void Function(long a);
+[DllImport("NativeLib", EntryPoint = "Function")]
+extern static void Function32(int a);
+
+#if WINDOWS
+    [DllImport("NativeLib", EntryPoint = "Function")]
+    extern static void Function64(int a);
+#else
+    [DllImport("NativeLib", EntryPoint = "Function")]
+    extern static void Function64(long a);
```

Comment:
nit: C# `long` is only correct on 64-bit Unix. The correct type here is `IntPtr` or `nint` (and `UIntPtr`/`nuint` for `unsigned long`)

Reply:
I see you actually discussed this in one of the collapsed threads.

I don't agree that it is incidental, the Unix ABI explicitly defines `unsigned long` to be the same size as `size_t` and therefore the same size as a pointer given the other constraints set in place.

Reply:
>nit: C# long is only correct on 64-bit Unix. 

It is also on macOS if I recall? I assume you are using Unix for all non-Windows platforms. That is the reason for the `64` suffix.

>The correct type here is IntPtr or nint (and UIntPtr/nuint for unsigned long)

I don't see it as the correct type. That is the pointer size and `long` isn't defined as being related to pointer size it just happens to align with that. It is a fine optimization for sure but as an example I prefer stating these sizes explicitly.

Reply:
> the Unix ABI explicitly defines unsigned long to be the same size as size_t and therefore the same size as a pointer given the other constraints set in place.

A link to this would be interesting. I was not aware that was an explicit decision.

Reply:
@tannergooding From what I am seeing this is the opposite - from https://uclibc.org/docs/psABI-x86_64.pdf.

> The type size_t is defined as unsigned long

Reply:
I guess it doesn't matter in principle - they are the same which we all agree. I don't see this as something that is required though because this is defined by a platform so it can be what they wish. I do accept it will work but as mentioned I prefer to keep this explicit and users are free to optimize as desired.

Reply:
> A link to this would be interesting. I was not aware that was an explicit decision.

https://gitlab.com/x86-psABIs/i386-ABI and https://gitlab.com/x86-psABIs/x86-64-ABI

Specifically: https://gitlab.com/x86-psABIs/x86-64-ABI/-/blob/master/x86-64-ABI/low-level-sys-info.tex#L261 and https://gitlab.com/x86-psABIs/i386-ABI/-/blob/hjl/master/low-level-sys-info.tex#L151

I would think the preferred pattern here is actually:
```
private static readonly bool IsWindows = RuntimeInformation.IsOSPlatform(OSPlatform.Windows);

[DllImport("NativeLib", EntryPoint = "Function")]
extern static void FunctionWindows(int a);

[DllImport("NativeLib", EntryPoint = "Function")]
extern static void FunctionUnix(nint a);

if (IsWindows)
{
    FunctionWindows(a);
}
else
{
    FunctionUnix(a);
}
```

This results in the irrelevant path being optimized away as dead code and matches what we effectively do for `CLong`/`CULong` internally.
You might theoretically want an assert or check for `IsUnix`, but that isn't necessary given .NET Framework or .NET Core.

Reply:
Alright. I've been convinced. I will update the example. Thanks @tannergooding 

/cc @danmoseley @jkoritzinsky

Reply:
I think this discussion was worthwhile because we're trying to establish a pattern for everyone. thanks.